### DISCUSSION TITLE
[Feature request / design discussion] How can we make Stack mode better?

### DISCUSSION THREAD

**Discussion by User670** [26F-Studio/Techmino#244] [Ideas]
With the current implementation of skyline height and scrolling camera, I'm having two minor issues that I feel can be improved:

- [Piece spawning] In stack mode pieces spawn at skyline aka 100 lines high, which means it takes a long time to soft drop a piece. (No, I'm not going to play with instant soft drop.) (Can we design the system so that the piece does not spawn too high above the stack? Should we?)
- [Scrolling camera] When moving the pieces left and right, the ghost often moves up and down the stack (due to the stack is often not perfect clear), and the camera follows the ghost and also move up and down a lot. (How can we redesign it so that the camera "stablizes"?)

[a comment by an automated account is not shown]

**Comment by MrZ626** [comment]
Finxing piece spawning position is too complex to explain in English for me, so I use Chinese below(((
出生太高确实下落慢，我想过几个解决方案但是都有缺点：
- 出生在当前场地最高的列往上20格（反正不是最顶上）
  - 就会让出块位置略微有些不可预测，插个棍子就高很多
  - 类似缺点1，这个方案并不一定能让出块高度符合直觉
  - 下落到视野内速度过快，会有些来不及反应（这一点也许是可以被接受的，反正我可以）
- 软降的时候如果方块太高会首先直接移动到场地最高点+20（或者10）的位置
  - 很奇怪

注意，以上两个设计只针对出生高度，不考虑镜头移动，镜头移动问题我已经有使用更多按键的解决方案了

**Reply by User670** [reply]
Translation

Spawning high does indeed make softdropping slow. I've been thinking of a few solutions but each have their drawbacks:
- Spawn 20 rows above the highest point of the stack (at the very least it's not the skyline)
  - it makes spawn location a bit unpredictable - sticking a long bar on the top of the stack makes this much higher
  - like drawback 1, it might not be intuitive where the piece spawns
  - if piece moves into the view too fast, it might be harder to react (maybe it's acceptable, I'm fine with this)
- when softdropping, if the block is too high, then first teleport to stack height +20 (or +10)
  - it's weird

Note that the above only deal with spawn height and not camera movement. I have a solution for camera using more buttons

**Comment by User670** [comment]
Using buttons to move camera doesn't negate the need to have a better default - there should be a good solution that works with most cases as a default, and leave the extra buttons for only the cases where the default can't account for (or those people who have very special preferences).

I've actually been thinking of something but forgot to post it here. Let's see if it works better:
- Camera should try to keep the top of the stack around the upper half of the screen. For example, for a stack of height 45, move the camera to display rows 31-50.
  - Notable exception: when the ghost piece moves too low (eg, hovering a long bar over a very deep Tetris well), move the camera down to keep the ghost around the bottom half of the screen.
- The upper edge of the camera acts as the skyline for the purpose of piece spawning, thus making the spawn position intuitive and consistent. Example, for the aforementioned 31-50 camera, piece spawn on row 51.

What do you think :-/

用按钮控制镜头不代表不用弄一个更好的默认方案。应该有一个比较好的、适用于大多数情况的解决方案作为默认，然后按钮用来应对特殊的情况（或者有特殊需要的人）。

我其实也在想一个方案但是忘了发到这里。看看这个方案和不合适：
- 镜头试图保持场地最高点在视野的上半部分。例如，场地45行高，镜头就显示31-50行。
  - 例外之一：当ghost位置太低（例如，准备把一根长条插进很深的一个洞里时），降低镜头使ghost在视野的下半部分。
- 镜头视野的上边缘作为方块生成的基准线。这样的生成位置比较符合直觉。例如，上述的31-50镜头，方块生成在51行。

这个你们怎么想 :-/

**Reply by MrZ626** [reply]
第一条除了在顶上一直蹭不会抖以外和现在的并没有很大区别，起楼的时候影子如果扫过底部同样会抖动，而且对于普通的20高场地有副作用，还得特殊处理（现在只在危险情况移动镜头的设计就是因为这个）

**Reply by MrZ626** [reply]
第二条跟我的第二条差不多，我没啥想法，只能说很电竞，机器人疯狂暴毙预定

**Comment by C29H25N3O5** [comment]
Well I this this problem is *partially* solved by the new Stack mode

[a comment by an automated account is not shown]

**Comment by C29H25N3O5** [comment]
Well MrZ now removed a staking mode and then hid the other two modes so that they can only be entered through Console...
I talked to MrZ and he told me that the new stacking mode is flawed and hiding them is a better choice
